### PR TITLE
MIN-1266: Off-chain Integration with StableSwap Smart Contract

### PR DESCRIPTION
**Pull request by m1n999999**

### CODE REVIEW

**Comment by m1n999999** [comment]
There are example txs:
Create Swap Order Tx: https://preprod.cardanoscan.io/transaction/cc12baafa3accb5b9eabe0f4b0a50c4f73de80849a83522c97cf65a32b3c5246
Craete Deposit Order Tx: https://preprod.cardanoscan.io/transaction/2b5deecef0e7bda781692152ec993d748363759135f763f3457bd15c7ecd437c
Create Withdraw Order Tx: https://preprod.cardanoscan.io/transaction/e67c77132297a13b7e664e2632b481849d6378c86ea561a041e50229a9b450d6
Craete Withdraw Imbalance Order Tx: https://preprod.cardanoscan.io/transaction/3762756b567c7c15c62bb7ed7a04eb16255a77e3afcf606af00b71f019af6a01
Create Zap Out Order Tx: https://preprod.cardanoscan.io/transaction/04493351b8b05c1638dc5c409c81eb2190abda60946eeb835cd6b4c4b020808d
Cancel Order Tx: https://preprod.cardanoscan.io/transaction/d3c30251307af69a91c91ca317f9b127ff50baff70d3d1469ef33904e5a859af

File: src/calculate.ts [hunk 1]
```diff
@@ -473,3 +475,516 @@ export namespace DexV2Calculation {
...
+    datumBalances,
+    totalLiquidity,
+    fee,
+    adminFee,
+    feeDenominator,
+  }: StableswapCalculateDepositOptions): bigint {
+    const tempDatumBalances = [...datumBalances];
+    const feeAmounts = tempDatumBalances.map((_) => 0n);
+    const assetVolumes = tempDatumBalances.map((_) => 0n);
+    const poolVolumes = tempDatumBalances.map((_) => 0n);
```
Comment (h2physics) [inline src/calculate.ts:731]: do we need these variables?

File: src/adapter.ts [hunk 2]
```diff
@@ -403,6 +403,21 @@ export class BlockfrostAdapter {
...
+      (pool) => pool.lpAsset === Asset.toString(lpAsset)
+    );
+    invariant(
+      pool,
+      `getStablePoolByLpAsset: Can not find pool by LP Asset ${Asset.toString(
+        lpAsset
+      )}`
+    );
+    return pool;
+  }
```
Comment (h2physics) [inline src/adapter.ts:419]: fetch single pool instead of all stable pools

File: src/stableswap.ts [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+      case StableOrder.StepType.WITHDRAW_IMBALANCE: {
+        return MetadataMessage.WITHDRAW_ORDER;
+      }
+      case StableOrder.StepType.ZAP_OUT: {
+        return MetadataMessage.ZAP_OUT_ORDER;
+      }
+    }
+  }
+
+  async buildCreateTx(options: OrderOptions): Promise<TxComplete> {
```
Comment (h2physics) [inline src/stableswap.ts:276]: createBulkOrdersTx or createManyOrdersTx

File: src/stableswap.ts [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+        batcherFee: batcherFee,
+        depositADA: FIXED_DEPOSIT_ADA,
+      };
+      tx.payToContract(
+        config.orderAddress,
+        {
+          inline: Data.to(StableOrder.Datum.toPlutusData(datum)),
+        },
+        orderAssets
+      ).payToAddress(sender, reductionAssets);
```
Comment (h2physics) [inline src/stableswap.ts:338]: check the reductionAssets is not empty

**Comment by m1n999999** [comment]
Tx that I apply fixed batcherFeeReduction function:
192da306b97e951bebfb8db9347a779196f0a5771d5b74d9962822ef3f6a4435
c268854a3a8793ce8f05cfadd0c4d2661e5fb3eabab032fc72329b57e51d5d45

**Review by h2physics — APPROVED** [review]

**Pull request merged**